File: cpp-examples/race/RaceHamster.cpp
#include "RaceHamster.h"

extern const int HAMSTER_COUNT = 4;

RaceHamster::RaceHamster(hamster::Territory& territory, const mpw::Location& location,
                         const mpw::Direction direction, const int grainCount) :
        Hamster(territory.shared_from_this(), location, direction, grainCount) {}

void RaceHamster::pickAndCount() {
    Hamster::pickGrain();
    grainCollected++;
}

int RaceHamster::getNumberOfCollectedGrains() const { return grainCollected; }

void RaceHamster::beginRightTurn() {
    Hamster::turnLeft();
    leftTurnsLeft = 2;
}

void RaceHamster::performRemainingRightTurn() {
    Hamster::turnLeft();
    leftTurnsLeft--;
}

bool RaceHamster::hasRemainingLeftTurnsForRightTurn() const {
    return leftTurnsLeft > 0;
}

File: java-examples/src/main/java/de/unistuttgart/iste/sqa/oo/hamstersimulator/race/RaceHamster.java
package de.unistuttgart.iste.sqa.oo.hamstersimulator.race;

import de.unistuttgart.hamster.facade.Hamster;
import de.unistuttgart.hamster.facade.Territory;
import de.unistuttgart.iste.sqa.mpw.framework.datatypes.Direction;
import de.unistuttgart.iste.sqa.mpw.framework.datatypes.Location;

public class RaceHamster extends Hamster {
    private int grainCollected = 0;
    private int leftTurnsLeft = 0;

    public RaceHamster(Territory territory,
                       Location location,
                       Direction direction) {
        super(territory, location, direction, 0);
    }

    public void pickAndCount() {
        super.pickGrain();
        grainCollected++;
    }

    public int getNumberOfCollectedGrains() {
        return grainCollected;
    }

    public void beginRightTurn() {
        super.turnLeft();
        leftTurnsLeft = 2;
    }

    public void performRemainingRightTurn() {
        super.turnLeft();
        leftTurnsLeft--;
    }

    public boolean hasRemainingLeftTurnsForRightTurn() {
        return leftTurnsLeft > 0;
    }
}
